electrum/lnpeer.py

```diff
@@ -223,12 +223,102 @@ def process_message(self, message):
             if asyncio.iscoroutinefunction(f):
                 asyncio.ensure_future(self.taskgroup.spawn(execution_result))
 
-    def on_error(self, payload):
+    def on_warning(self, payload):
+        """When we receive a warning, we put it into the ordered message queue for the channel.
+        If we `self.wait_for_message` in a protocol action, this will lead to a disconnect via
+        an exception raised therein.
+        """
+        # TODO: we could need some reconnection logic here -> delayed reconnect
+        self.logger.info(f"remote peer sent warning [DO NOT TRUST THIS MESSAGE]: {payload['data'].decode('ascii')}")
+        channel_id = payload.get("channel_id")
+        if channel_id == bytes(32):
+            for cid in self.channels.keys():
+                self.ordered_message_queues[cid].put_nowait((None, {'warning': payload['data']}))
+            raise GracefulDisconnect
+        warned_channel_id = None
+        if channel_id in self.temp_id_to_id:
+            warned_channel_id = self.temp_id_to_id[channel_id]
+        elif channel_id in self.channels:
+            warned_channel_id = channel_id
+        if warned_channel_id:
+            # MAY disconnect.
+            self.ordered_message_queues[warned_channel_id].put_nowait((None, {'warning': payload['data']}))
+            raise GracefulDisconnect
+
+    async def on_error(self, payload):
+        """When we receive an error, we close out the channel of the corresponding
+        channel id (all if channel_id is zero). If we `self.wait_for_message` in a
+        protocol action, this will lead to a disconnect via an exception raised therein.
+        """
         self.logger.info(f"remote peer sent error [DO NOT TRUST THIS MESSAGE]: {payload['data'].decode('ascii')}")
-        chan_id = payload.get("channel_id")
-        if chan_id in self.temp_id_to_id:
-            chan_id = self.temp_id_to_id[chan_id]
-        self.ordered_message_queues[chan_id].put_nowait((None, {'error':payload['data']}))
+        channel_id = payload.get("channel_id")
+        # if channel_id is all zero: MUST fail all channels with the sending node.
+        if channel_id == bytes(32):
+            for cid in self.channels.keys():
+                await self.try_force_closing(cid)
+                self.ordered_message_queues[cid].put_nowait((None, {'error': payload['data']}))
+            raise GracefulDisconnect
+        # otherwise: MUST fail the channel referred to by channel_id, if that channel is with the sending node.
+        erring_channel_id = None
+        if channel_id in self.temp_id_to_id:
+            erring_channel_id = self.temp_id_to_id[channel_id]
+        elif channel_id in self.channels:
+            erring_channel_id = channel_id
+        if erring_channel_id:
+            await self.try_force_closing(erring_channel_id)
+            self.ordered_message_queues[erring_channel_id].put_nowait((None, {'error': payload['data']}))
```

Comment:
It is error-prone to have an `async` `on_*` message-handler that puts messages into `ordered_message_queues`.
It is nicely visible here as you have an `await` directly above this line: you have now lost all guarantees about the messages being "ordered". That is, if two messages are received for the same channel, they might end up in `ordered_message_queues[chan_id]` in any order this way.
For this reason, I suggest keeping `on_error` as non-async.

Reply:
I see, thanks, done